src/quick-lint-js/error_decorator.h

```diff
@@ -0,0 +1,54 @@
+// quick-lint-js finds bugs in JavaScript programs.
+// Copyright (C) 2020  Matthew Glazar
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+#ifndef QUICK_LINT_JS_ERROR_DECORATOR_H
+#define QUICK_LINT_JS_ERROR_DECORATOR_H
+
+#include <quick-lint-js/error.h>
+#include <quick-lint-js/text-error-reporter.h>
+#include <quick-lint-js/vim-qflist-json-error-reporter.h>
+#include <variant>
+
+namespace quick_lint_js {
+class error_decorator : public error_reporter {
+ public:
+  using reporter_variant = std::variant<text_error_reporter,
+    vim_qflist_json_error_reporter>;
```

Comment:
I think `error_decorator` should not be coupled to any particular `error_reporter` derived class. Ideas:

* Make the decorated `error_reporter` type a template parameter.
* Make the decorated `error_reporter` object live outside `error_decorator`, and give `error_decorator` a pointer to the decorated `error_reporter` object.

Reply:
I gave a `error_reporter *` to `error_decorator`.
I had to modify heavily the `any_error_reporter` class in `main.cpp` to do that so let me know if my modifications are bad or not so C++ (I am not really used to template).